### PR TITLE
Implement TryFrom<Box<[T]>> for Box<[T; N]>

### PR DESCRIPTION
**Pull request by nvzqz**
These changes serve to provide the same benefits as PR #44764 except for boxes, which refer to owned arrays allocated in the heap.

Should these changes be extended to other types like `Arc<[T]>` and `Rc<[T]>`?

r? @sfackler

### CODE REVIEW

File: src/liballoc/boxed.rs [hunk 1]
```diff
@@ -623,6 +624,66 @@ impl From<Box<str>> for Box<[u8]> {
...
+    /// moving out the boxed slice, so that a copy of the slice does not need to
+    /// be made.
+    #[inline]
+    pub fn into_boxed_slice(self) -> Box<[T]> {
+        self.0
+    }
+}
+
+macro_rules! array_impls {
+    ($($N:expr)+) => {
```
Comment (kennytm) [inline src/liballoc/boxed.rs]: You forgot to add `$(...)+` around the item.

```
[00:05:12]    Compiling alloc v0.0.0 (file:///checkout/src/liballoc)
[00:05:12] error: variable 'N' is still repeating at this depth
[00:05:12]    --> liballoc/boxed.rs:665:47
[00:05:12]     |
[00:05:12] 665 |         impl<T> TryFrom<Box<[T]>> for Box<[T; $N]> {
[00:05:12]     |                                               ^^
[00:05:12] 
[00:05:12] error: Could not compile `alloc`.
```

**Review by kennytm — COMMENTED** [review]
Is commit 781574787ae169ba4fe87a71730a5a2b1638761b a mistake? Why is there a merge commit in this PR?

File: src/liballoc/boxed.rs (hunk 1 above)
Comment (nvzqz) [reply src/liballoc/boxed.rs]: I add the current state of the master branch and idk why all of that is there. I'll reset back to the initial commit tomorrow.

File: src/liballoc/boxed.rs [hunk 2]
```diff
@@ -623,6 +624,69 @@ impl From<Box<str>> for Box<[u8]> {
...
+        $(
+            #[unstable(feature = "try_from", issue = "33417")]
+            impl<T> TryFrom<Box<[T]>> for Box<[T; $N]> {
+                type Error = TryFromSliceError<T>;
+
+                #[inline]
+                fn try_from(slice: Box<[T]>) -> Result<Box<[T; $N]>, TryFromSliceError<T>> {
+                    if slice.len() == $N {
+                        let ptr = Box::into_raw(slice) as *mut [T; $N];
+                        unsafe { Ok(Box::from_raw(ptr)) }
```
Comment (sfackler) [inline src/liballoc/boxed.rs]: @alexcrichton is this a valid operation?

Comment (alexcrichton) [reply src/liballoc/boxed.rs]: AFAIK yeah this should be safe, the destructor for `Box<T>` should deallocate with the size of `T` which in this case would be `mem::size_of::<T>() * N`

**Comment by sfackler** [comment]
Could you give some examples where you'd use these conversions?

**Comment by nvzqz** [comment]
[Here][post] a Redditor asked about how to go about making a heap-allocated table with a static size. I recommended using `Vec` and then converting that into a boxed slice. Because the size is known at compile-time, there's not much reason to keep the dynamic length.

Having const generics as well as a safe conversion interface such as this one would solve his problem without the need for `unsafe`. So while this conversion as-is wouldn't solve his problem, it in a way potentially solves half of it.

In my experience, LLVM seems to optimize further with arrays (`&[T; N]`) than just slices (`&[T]`). For people who want to make their code as fast as possible, they may want to reach for fixed-size arrays.

We currently have `&[T]` to `&[T; N]` conversions, I think it also makes sense to have the corresponding box conversions.

[post]: https://www.reddit.com/r/rust/comments/7nhq34/whether_to_use_vect_or_boxt_n_for_a_fixedsize/

**Comment by sfackler** [comment]
That post was asking about `[T; 262144]`, though. Are these conversions worth it before const generics?

**Comment by carols10cents** [comment]
Ping @nvzqz, looks like @sfackler has a further question for you?

**Comment by nvzqz** [comment]
@sfackler to me it makes sense to have these conversions when we already have `&[T]` to `&[T; N]` conversions. If these conversions don't get added now, I'll be submitting another similar PR once const generics are in nightly.

Currently, unsafe code is needed by libraries to make a statically sized box. If a statically-size slice is needed, multiple checks must be made for whether the box is the right size if authors choose to avoid `unsafe`.

**Comment by sfackler** [comment]
> Currently, unsafe code is needed by libraries to make a statically sized box.

??

```rust
let b: Box<[u64; 100]> = Box::new([0; 100]);
```

**Comment by nvzqz** [comment]
I should have elaborated. Your example is possible with trivial/primitive types that implement `Copy`. However, one may want to do something like this:

```rust
struct Entry {
    /* fields */
}

const NUM_ENTRIES: usize = 32;

lazy_static! {
    static ref TABLE: Box<[Entry; NUM_ENTRIES]> = {
        let mut table = Vec::<Entry>::with_capacity(NUM_ENTRIES);
        for i in 0..NUM_ENTRIES {
            table.push(Entry { /* ... */ });
        }
        table.into_boxed_slice().try_into().unwrap()
    };
}
```

By using the `Vec` to `Box<[T]>` conversion as well as these changes, this can be done without `unsafe`.

**Comment by sfackler** [comment]
But how big is NUM_ENTRIES in practice? I'd expect it to almost always be larger than 32.

**Comment by nvzqz** [comment]
Well, generally, perhaps. Unless this is something that's polled, one can't really give a proper number. Alas, it is likely to be greater than 32.

**Comment by kennytm** [comment]
Hi @sfackler @nvzqz, ping from triage. 

Given the proper use case of this usually require N >> 32, I guess we should postpone this PR until const generics is implemented?

**Comment by pietroalbini** [comment]
@sfackler @nvzqz ping from triage! What's the status of this PR?

**Comment by shepmaster** [comment]
Since we haven't heard from @sfackler in a few weeks, I'm randomly reassigning from the @rust-lang/libs team...

r? @dtolnay

**Comment by dtolnay** [comment]
I would be inclined to accept this by analogy with #44764. If we decide that array stuff is not worth doing until const generics are available, let's roll back the &[T]⟶&[T; N] conversion impls.

@rfcbot fcp merge

**Comment by dtolnay** [comment]
TryFrom is unstable so I am going to unilaterally accept these and we can consider more closely as a team when stabilizing TryFrom. Thanks @nvzqz!

@rfcbot fcp cancel

File: src/liballoc/boxed.rs [hunk 3]
```diff
@@ -623,6 +624,79 @@ impl From<Box<str>> for Box<[u8]> {
     }
 }
 
+/// The error type returned when a conversion from a boxed slice to a boxed
+/// array fails.
+#[unstable(feature = "try_from", issue = "33417")]
+#[derive(Clone)]
+pub struct TryFromSliceError<T>(Box<[T]>);
```
Comment (dtolnay) [inline src/liballoc/boxed.rs]: Please add a brief example involving TryFromSliceError. See [FromUtf8Error](https://doc.rust-lang.org/std/string/struct.FromUtf8Error.html) for examples of examples.

File: src/liballoc/boxed.rs [hunk 4]
```diff
@@ -623,6 +624,79 @@ impl From<Box<str>> for Box<[u8]> {
...
+    }
+
+    /// Returns the boxed slice that was attempted to convert to a boxed array.
+    ///
+    /// This method is meant to avoid allocation. It will consume the error,
+    /// moving out the boxed slice, so that a copy of the slice does not need to
+    /// be made.
+    #[unstable(feature = "try_from", issue = "33417")]
+    #[inline]
+    pub fn into_boxed_slice(self) -> Box<[T]> {
```
Comment (dtolnay) [inline src/liballoc/boxed.rs]: Please add an example showing a case in which this method is useful. The point of the example wouldn't be to show **how** to call this method (we can assume readers know how to call methods when they get to browsing TryFrom documentation) but to show **why** someone would want to call this method.

File: src/liballoc/boxed.rs [hunk 5]
```diff
@@ -623,6 +624,79 @@ impl From<Box<str>> for Box<[u8]> {
...
+
+macro_rules! array_impls {
+    ($($N:expr)+) => {
+        $(
+            #[unstable(feature = "try_from", issue = "33417")]
+            impl<T> TryFrom<Box<[T]>> for Box<[T; $N]> {
+                type Error = TryFromSliceError<T>;
+
+                #[inline]
+                fn try_from(slice: Box<[T]>) -> Result<Box<[T; $N]>, TryFromSliceError<T>> {
```
Comment (dtolnay) [inline src/liballoc/boxed.rs]: Please add a test so that we don't break this implementation later. Something similar to the test in #44764 should work.

**Comment by pietroalbini** [comment]
Hi from the release team @nvzqz, there are a few comments from @dtolnay that needs to be addressed, can you fix them so this PR can be merged?

**Comment by pietroalbini** [comment]
Thank you so much for this PR @nvzqz! Unfortunately, we haven't heard from you for more than two weeks, so I'm closing this PR to keep the queue tidy.

**Please reopen it** if you have time to address @dtolnay comments in the near future!

**Pull request closed without merging**